[z80asm] Recent bug `assertion failed in modlink.c:512`. It was OK a couple of days ago.

This is a very recent bug. It was not present in the snapshot from 18/09/2021.
It is present in today's snapshot (29/09/2021).

If I try to compile the "tiles" test in Cross-Lib for the Aquarius using SCCZ80 I get this error:

```
-o ../build/Xtiles_aquarius_exp_16k
assertion failed in modlink.c:512
make: *** [tests/tiles/Makefile.tiles:691: aquarius_exp_16k] Error 1
````

I will try to reproduce with a smaller example if this bug report is not immediately clear.
In Cross-Lib this can be reproduced by running from the `src` :
`xl build tiles aquarius`
which is equivalent to
`make aquarius -f tests/tiles/Makefile.tiles`

The Makefile target looks like
```
aquarius_exp_16k: 
	$(Z88DK_PATH)$(MYZ88DK) +aquarius $(SCCZ80_OPTS) -clib=ansi $(INCLUDE_OPTS) -vn \
	-DSOUNDS -D__AQUARIUS__ \
	-lndos $(NO_UDG_FLAG) \
	-D_XL_SLOW_DOWN_FACTOR=$(_COMMON_SLOWDOWN_FACTOR)*$(_AQUARIUS_SLOWDOWN) \
	-DCONIO -DNO_INIT_GRAPHICS \
	$(FULL_FILES) \
	-create-app	\
	-o $(BUILD_PATH)/X$(GAME_NAME)_aquarius_exp_16k
```

The other games aned tests seem to compile OK.

--- a/src/z80asm/modlink.c
+++ b/src/z80asm/modlink.c
@@ -483,6 +483,22 @@ static int compute_equ_exprs_once(ExprList* exprs, bool show_error, bool module_
 		return 0;
 }
 
+/* check if we have expressions still with target, i.e. circular definitions (see #1869) */
+static void check_equ_exprs_solved(ExprList* exprs, bool module_relative_addr) {
+	if (module_relative_addr)			/* not linking */
+		return;
+
+	ExprListElem* iter = ExprList_first(exprs);
+	while (iter != NULL) {
+		Expr* expr = iter->obj;
+		if (expr->target_name) {
+			set_expr_env(expr, module_relative_addr);
+			error_not_defined(expr->target_name);
+		}
+		iter = ExprList_next(iter);
+	}
+}
+
 /* compute all equ expressions, removing them from the list */
 void compute_equ_exprs(ExprList* exprs, bool show_error, bool module_relative_addr)
 {
@@ -496,6 +512,9 @@ void compute_equ_exprs(ExprList* exprs, bool show_error, bool module_relative_ad
 	/* if some unresolved, give up and show error */
 	if (show_error && compute_result < 0)
 		compute_equ_exprs_once(exprs, true, module_relative_addr);
+
+	/* if linking, check that we solved all EQU expressions */
+	check_equ_exprs_solved(exprs, module_relative_addr);
 }
 
 /* compute and patch expressions */

--- a/src/z80asm/t2/issue_1869.t
+++ b/src/z80asm/t2/issue_1869.t
@@ -1,0 +1,34 @@
+#!/usr/bin/env perl
+
+BEGIN { use lib 't2'; require 'testlib.pl'; }
+
+# z80asm: assertion failed in modlink.c:512
+# https://github.com/z88dk/z88dk/issues/1869
+
+z80asm_nok("", "", <<END_ASM, <<END_ERR);
+        defc    i_64 = i_61
+        defc    i_61 = i_64
+END_ASM
+Error at file '$test.asm' line 1: symbol 'i_64' not defined
+Error at file '$test.asm' line 2: symbol 'i_61' not defined
+END_ERR
+
+path("$test.1.asm")->spew(<<END_ASM);
+        public  i_64
+        extern  i_61
+        defc    i_64 = i_61
+END_ASM
+
+path("$test.2.asm")->spew(<<END_ASM);
+        public  i_61
+        extern  i_64
+        defc    i_61 = i_64
+END_ASM
+
+capture_nok("./z88dk-z80asm -b $test.1.asm $test.2.asm", <<END_ERR);
+Error at file '$test.1.asm' line 3: symbol 'i_64' not defined
+Error at file '$test.2.asm' line 3: symbol 'i_61' not defined
+END_ERR
+
+unlink_testfiles;
+done_testing;

